components/brave_vpn/utils_win.cc

```diff
@@ -0,0 +1,327 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/utils_win.h"
+
+#include <ras.h>
+#include <raserror.h>
+#include <stdio.h>
+
+#include "base/logging.h"
+#include "base/strings/stringprintf.h"
+
+#define DEFAULT_PHONE_BOOK NULL
+
+namespace brave_vpn {
+
+namespace {
+
+// https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessage
+void PrintSystemError(DWORD error) {
+  constexpr DWORD kBufSize = 512;
+  TCHAR lpsz_error_string[kBufSize];
+
+  DWORD buf_len =
+      FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                    NULL, error, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                    lpsz_error_string, kBufSize, NULL);
+  if (buf_len) {
+    LOG(ERROR) << lpsz_error_string;
+  }
+}
+
+// https://docs.microsoft.com/en-us/windows/win32/api/ras/nf-ras-rassetcredentialsa
+DWORD SetCredentials(LPCTSTR entry_name, LPCTSTR username, LPCTSTR password) {
+  RASCREDENTIALS credentials;
+
+  ZeroMemory(&credentials, sizeof(RASCREDENTIALS));
+  credentials.dwSize = sizeof(RASCREDENTIALS);
+  credentials.dwMask = RASCM_UserName | RASCM_Password;
+
+  wcscpy_s(credentials.szUserName, UNLEN + 1, username);
+  wcscpy_s(credentials.szPassword, PWLEN + 1, password);
+
+  DWORD dw_ret =
+      RasSetCredentials(DEFAULT_PHONE_BOOK, entry_name, &credentials, FALSE);
+  if (dw_ret != ERROR_SUCCESS) {
+    internal::PrintRasError(dw_ret);
+    return dw_ret;
+  }
+
+  return ERROR_SUCCESS;
+}
+
+}  // namespace
+
+namespace internal {
+
+// https://docs.microsoft.com/en-us/windows/win32/api/ras/nf-ras-rasgeterrorstringa
+void PrintRasError(DWORD error) {
+  DWORD kBufSize = 512;
+  TCHAR lpsz_error_string[kBufSize];
+
+  if (error > RASBASE && error < RASBASEEND) {
+    if (RasGetErrorString(error, lpsz_error_string, kBufSize) ==
+        ERROR_SUCCESS) {
+      LOG(ERROR) << lpsz_error_string;
+      return;
+    }
+  }
+
+  PrintSystemError(error);
+}
+
+std::wstring GetPhonebookPath() {
+  TCHAR* app_data_path = NULL;
+  std::wstring phone_book_path;
+
+  // https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-expandenvironmentstringsa
+  // Caculate required buf size first.
+  DWORD dw_buf_size =
```

Comment:
nit: We may as well make this one `const` so that it's not accidentally modified later.

Reply:
Done.